core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -568,6 +569,7 @@ public PlanOptimizers(
 
         // Perform redirection before CBO rules to ensure stats from destination connector are used
         // Perform redirection before agg, topN, limit, sample etc. push down into table scan as the destination connector may support a different set of push downs
+        // Perform redirection before push down of dereferences into table scan via PushProjectionIntoTableScan
         // Perform redirection after at least one PredicatePushDown and PushPredicateIntoTableScan to allow connector to use pushed down predicates in redirection decision
```

Comment:
(not in scope of this PR) it seems a bit arbitrary to me to let the destination table to be aware of predicate pushdown, but not other pushdowns. in long term, would it be better to let the `applyTableScanRedirect` API return equivalents of "residues" for all the pushdowns that it doesn't understand - and recreate those nodes in the engine?

Reply:
You are right that we could potentially extend applyTableScanRedirect to "reverse" every type of pushdown. But the implementation becomes more complex with every type of pushdown we try to deal with.
The awareness of pushed down predicate was added as it could potentially be a factor in the redirection decision (there may be multiple candidate destinations and we might choose between them based on the predicate).

Reply:
In the future, we should rely on materialized views substituion. Optimizer will be able to match MV vs query subplan and substitute subplan with MV scan. With that solution, there is no problem of "reversing" pushdown.